### DISCUSSION TITLE
find_cycle() unexpectedly slow for certain graphs

### DISCUSSION THREAD

**Discussion by OvervCW** [networkx/networkx#6543] [Q&A]
Consider the following two programs that each use `find_cycle` on slightly different graphs that represent a long chain of nodes:

## Program A
```python
import networkx as nx

g = nx.DiGraph()

for i in range(1000):
    g.add_node(str(i))

    if i > 0:
        g.add_edge(str(i), str(i - 1))

nx.find_cycle(g)
```

## Program B
```python
import networkx as nx

g = nx.DiGraph()

for i in range(1000):
    g.add_node(str(i))

    if i < 999:
        g.add_edge(str(i), str(i + 1))

nx.find_cycle(g)
```

The number of nodes and edges is identical. The only difference is that they're connected in reverse order. However, the first program takes ~2 seconds to find that there is no cycle and the second program just <0.1 seconds.

I would expect this algorithm to be much faster for such a small graph. What causes it to be so slow?

**Comment by PurviChaurasia** [comment]
I think, this is much to do in the manner in which the nodes are added. In program A, the edges are added in reverse order, which may cause DFS to visit nodes in a less efficient way, resulting in a longer runtime for finding that there is no cycle. In contrast, program B adds edges in the order that DFS may visit nodes more efficiently, leading to a faster runtime for finding no cycle. Could there be a work around for this? What do you think?

**Comment by navyagarwal** [comment]
@OvervCW

My observation over here is that when I pass `orientation="ignore"` to `find_cycle` in both the code snippets, their running time are almost equal and around 1ms. 

When I pass `orientation="reverse"`, their run times are switched.

For `orientation="original"`, the codes run in the same time as when orientation has the default `None` value, our results are contradicting here somehow.

**Reply by OvervCW** [reply]
Ah, you're right. I accidentally passed "original" as start node instead of orientatiion, I'm sorry for that.

**Reply by navyagarwal** [reply]
Another observation, when I pass a source to Program A, whether `source='0'` or `source='999'`, the code runs in ~1ms. Same goes the case for Program B as well (i.e. specifying the source for Program B does not bring much difference in run time)

**Reply by OvervCW** [reply]
I found something else that's interesting. If I use `networkx.algorithms.dag.has_cycle(g)` then the program is fast in both cases.

**Reply by navyagarwal** [reply]
I was looking through the documentation of the DAG module [here](https://networkx.org/documentation/stable/reference/algorithms/dag.html#module-networkx.algorithms.dag) but did not find this method.

I found its code in the repository as - 
```python
def has_cycle(G):
    """Decides whether the directed graph has a cycle."""
    try:
        # Feed the entire iterator into a zero-length deque.
        deque(topological_sort(G), maxlen=0)
    except nx.NetworkXUnfeasible:
        return True
    else:
        return False
```
I think the catch is that this method only works for directed graphs, it throws an error for undirected graphs whereas find_cycle works for both directed and undirected.

**Reply by PurviChaurasia** [reply]
Hmmm but in that case don't you think, there should be a similar `has_cycle` function for undirected graphs as well? Rather than returning the list of nodes present in the cycle (which is what `find_cycle` does) a similar function should also exist as the use case for such a scenario is high. @dschult @MridulS

**Reply by OvervCW** [reply]
Sure, but you would expect the performance to be similar on a directed graph.

**Reply by PurviChaurasia** [reply]
Yep that's true

**Reply by PurviChaurasia** [reply]
@dschult @MridulS Should I work on such a function?
